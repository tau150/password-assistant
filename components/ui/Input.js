import * as React from 'react'
import PropTypes from 'prop-types'
import { Eye, EyeOff, Check } from "react-feather"
import { FormattedMessage } from 'react-intl'
import { Info } from "react-feather"
import { FormControl, FormLabel, FormHelperText, InputGroup, Input as CHInput, Tooltip, InputRightElement, Icon, Box, Text } from '@chakra-ui/react'

import { validateInput, extractI18nConfig } from '../../utils/formValidation'

function getLabelColor(errors, validTypes, type) {
  if(!errors && !validTypes){
    return 'gray.500'
  }
  if(errors?.includes(type)){
    return 'red.300'
  }
  if(validTypes.includes(type)){
    return 'green.300'
  }

}
function Input(props){
  const { name, id, label, handleOnChange, placeholder, validations, type,
    tooltipText, ariaTooltip, shouldCountChars, shouldToggleShow, inputLimit } = props

  const [show, setShow] = React.useState(false)
  const [inputValue, setInputValue] = React.useState('')
  const [isFocused, setIsFocused] = React.useState(false)
  const [errors, setErrors] = React.useState(null)
  const [isValid, setIsValid] = React.useState(true)
  const [validTypes, setValidTypes] = React.useState(null)

  const handleInputOnChange = (e) => {
    const { value } = e.target
    let isValidInput = true

    if(validations){
      const { isValid: successValidation, errors, validTypes} = validateInput(value, validations)
      isValidInput = successValidation
      setErrors(errors)
      setIsValid(isValidInput)
      setValidTypes(validTypes)
    }
    if(shouldCountChars && (inputLimit < value.length)){
      return
    }
    setInputValue(value)
    handleOnChange?.(e, isValidInput)
  }

  const handleClick = () => setShow(!show)
  const validationReferences = validations ? extractI18nConfig(validations) : null

  const showErrorMessages = (isFocused || !isValid) && validationReferences
  const showValidTick = isValid && !errors?.length && inputValue && validations

  return (
    <FormControl isInvalid={!isValid}>
      <FormLabel htmlFor={id} d='flex' alignItems='center'>
        {label}
        {tooltipText && (<Tooltip
          hasArrow placement='top'
          label={tooltipText}
          aria-label={ariaTooltip}>
            <Icon ml={2} color='blue.500' as={Info} />
          </Tooltip>)}
        </FormLabel>
        <InputGroup size="md" position='relative'>
        <CHInput
          id={id}
          name={name}
          value={inputValue}
          pr="4.5rem"
          type={shouldToggleShow ? (show ? "text" : "password") : type}
          placeholder={placeholder}
          onChange={handleInputOnChange}
          onFocus={()=> setIsFocused(true)}
          onBlur={()=> setIsFocused(false)}
        />
        <InputRightElement width="4.5rem">
          {shouldToggleShow && (
            show ? <Icon as={EyeOff} onClick={handleClick}/> : <Icon as={Eye} onClick={handleClick}/>
          )}
        </InputRightElement>
        {showValidTick && <Icon right='-40px' position='absolute' m={4} as={Check} color='green.400' />}
      </InputGroup>
      {shouldCountChars && <FormHelperText textAlign='right'>{`${inputValue.length}/${inputLimit}`}</FormHelperText>}
      {showErrorMessages && (
        <Box mt={2} px={2} py={1} bgColor='gray.100' position='absolute'>
          {validationReferences?.map(validation => <Text mt={1} color={getLabelColor(errors, validTypes, validation.type)} key={validation.type}>
           <FormattedMessage id={validation.i18nId} defaultMessage={validation.defaultMessage} values={{ [validation.i18nVariable]: validation.i18nValue }}/>
          </Text>)}
      </Box>
      )}
    </FormControl>
  )
}

Input.defaultProps = {
  inputLimit: 60,
  shouldCountChars: false,
  shouldToggleShow: false,
  type: 'text'
}

Input.propTypes = {
  name: PropTypes.string,
  label: PropTypes.string,
  id: PropTypes.string,
  placeholder: PropTypes.string,
  shouldToggleShow: PropTypes.bool,
  shouldCountChars: PropTypes.bool,
  inputLimit: PropTypes.number,
  type: PropTypes.string,
  validations: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.string,
    reference: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  }))
}

export default Input