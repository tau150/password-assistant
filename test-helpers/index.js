
import * as React from 'react'
import {   render as rtlRender } from '@testing-library/react'
import { IntlProvider } from 'react-intl';
import { ChakraProvider } from "@chakra-ui/react"

import theme from '../styles/theme';
import intlENmessages from '../content/locale/en.json'

function AppProviders({children}){
  return (
    <IntlProvider
      locale='en'
      messages={intlENmessages}
      onError={() => null}>
      <ChakraProvider theme={theme}>
        {children}
      </ChakraProvider>
    </IntlProvider>
  )
}


export function render(ui, ...renderOptions){
  const returnValue = {
    ...rtlRender(ui, {wrapper: AppProviders, ...renderOptions}),
  }
  return returnValue
}
