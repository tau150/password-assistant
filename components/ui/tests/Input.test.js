import * as React from 'react'
import { getByLabelText, getByText, screen } from '@testing-library/dom'
import userEvent from '@testing-library/user-event';
import { render } from '../../../test-helpers'

import Input from '../Input'


let props;

const inputName = 'testInput'
const placeholder = 'Some placeholder'
const tooltipText = 'Some tooltip text'
const ariaTooltip = 'some aria text'
const label = 'Awesome input'
const handleOnChange = jest.fn()

const minValidationMessage = 'Must contain at least 5 characters'
const upperValidationMessage = 'Must contain at least one upper characters'

beforeEach(() => {
  props = {
    type: 'text',
    shouldCountChars: false,
    shouldToggleShow: true,
    inputLimit: null,
    name: inputName,
    id: inputName,
    validations: [{type: 'min', reference: 5}, {type: 'oneUpper'}],
    inputLimit: 60,
    handleOnChange,
    ariaTooltip,
    label,
    tooltipText,
    placeholder,
  }
})


describe('Input test', () => {
  it('renders Input', () => {
    expect(render(<Input {...props} />)).toMatchSnapshot()
  })

  it('show the correct props', () => {
    render(<Input {...props} />)

    expect(screen.getByLabelText(label)).toBeInTheDocument()
    expect(screen.queryByText('0/60')).not.toBeInTheDocument()
    expect(screen.getByRole('tooltip', {name: ariaTooltip})).toBeInTheDocument()
  })

  it('Do not show tooltip if there is no tooltip text', () => {
    props.tooltipText = null
    render(<Input {...props} />)

    expect(screen.getByLabelText(label)).toBeInTheDocument()
    expect(screen.queryByRole('tooltip', {name: ariaTooltip})).not.toBeInTheDocument()
  })

  it('Show input limit when shouldCountChars is true', () => {
    props.shouldCountChars = true
    render(<Input {...props} />)
    expect(screen.getByText('0/60')).toBeInTheDocument()
  })

  it('Show input limit using the provided limit', () => {
    props.shouldCountChars = true
    props.inputLimit = 200
    render(<Input {...props} />)
    expect(screen.getByText('0/200')).toBeInTheDocument()
  })

  it('Switch password visualization', () => {
    render(<Input {...props} />)

    let eyeOffIcon = screen.queryByTestId(/eye-off/i)
    let eyeOnIcon = screen.getByTestId(/eye-on/i)

    expect(eyeOffIcon).not.toBeInTheDocument()
    expect(eyeOnIcon).toBeInTheDocument()

    userEvent.click(eyeOnIcon)

    eyeOffIcon = screen.getByTestId(/eye-off/i)
    eyeOnIcon = screen.queryByTestId(/eye-on/i)

    expect(eyeOffIcon).toBeInTheDocument()
    expect(eyeOnIcon).not.toBeInTheDocument()
  })

  it('Show validations descriptions on focus and feedback after typing', () => {
    render(<Input {...props} />)

    const input = screen.getByRole('input', { name: inputName })
    userEvent.click(input)

    let validationsReferences = screen.getAllByRole('status', {name: 'regular'})
    expect(validationsReferences[0]).toHaveTextContent(minValidationMessage)
    expect(validationsReferences[1]).toHaveTextContent(upperValidationMessage)

    userEvent.type(input, '123L')
    validationsReferences = screen.queryByRole('status', {name: 'regular'})
    expect(validationsReferences).not.toBeInTheDocument()

   const errorValidationsReference = screen.getByRole('status', {name: 'error'})
   expect(errorValidationsReference).toHaveTextContent(minValidationMessage)

   const successValidationsReference = screen.getByRole('status', {name: 'success'})
   expect(successValidationsReference).toHaveTextContent(upperValidationMessage)

   const successTick = screen.queryByTestId(/positive-tick/i)
   expect(successTick).not.toBeInTheDocument()
   expect(handleOnChange).toHaveBeenCalled()
  })

  it('Show validation tick when an input is valid', () => {
    render(<Input {...props} />)

    const input = screen.getByRole('input', { name: inputName })
    userEvent.click(input)

    let validationsReferences = screen.getAllByRole('status', {name: 'regular'})
    expect(validationsReferences[0]).toHaveTextContent(minValidationMessage)
    expect(validationsReferences[1]).toHaveTextContent(upperValidationMessage)

    userEvent.type(input, '123L644')
    validationsReferences = screen.queryByRole('status', {name: 'regular'})
    expect(validationsReferences).not.toBeInTheDocument()

   const errorValidationsReference = screen.queryByRole('status', {name: 'error'})
   expect(errorValidationsReference).not.toBeInTheDocument()

   const successValidationsReference = screen.getAllByRole('status', {name: 'success'})
   expect(successValidationsReference[0]).toHaveTextContent(minValidationMessage)
   expect(successValidationsReference[1]).toHaveTextContent(upperValidationMessage)

   const successTick = screen.getByTestId(/positive-tick/i)
   expect(successTick).toBeInTheDocument()
   expect(handleOnChange).toHaveBeenCalled()
  })
})

