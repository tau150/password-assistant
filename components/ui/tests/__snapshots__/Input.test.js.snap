// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input test renders Input 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-1bpd1fc"
          for="testInput"
          id="field-1-label"
        >
          Awesome input
          <svg
            aria-label="some aria text"
            class="chakra-icon css-ehdjs0"
            fill="none"
            focusable="false"
            height="24"
            role="tooltip"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <line
              x1="12"
              x2="12"
              y1="16"
              y2="12"
            />
            <line
              x1="12"
              x2="12.01"
              y1="8"
              y2="8"
            />
          </svg>
        </label>
        <div
          class="chakra-input__group css-4302v8"
        >
          <input
            aria-label="testInput"
            class="chakra-input css-155nkyn"
            id="testInput"
            name="testInput"
            placeholder="Some placeholder"
            role="input"
            type="password"
            value=""
          />
          <div
            class="chakra-input__right-element css-ohd79a"
          >
            <svg
              class="chakra-icon css-13otjrl"
              data-testid="eye-on"
              fill="none"
              focusable="false"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-1bpd1fc"
        for="testInput"
        id="field-1-label"
      >
        Awesome input
        <svg
          aria-label="some aria text"
          class="chakra-icon css-ehdjs0"
          fill="none"
          focusable="false"
          height="24"
          role="tooltip"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="12"
            x2="12"
            y1="16"
            y2="12"
          />
          <line
            x1="12"
            x2="12.01"
            y1="8"
            y2="8"
          />
        </svg>
      </label>
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          aria-label="testInput"
          class="chakra-input css-155nkyn"
          id="testInput"
          name="testInput"
          placeholder="Some placeholder"
          role="input"
          type="password"
          value=""
        />
        <div
          class="chakra-input__right-element css-ohd79a"
        >
          <svg
            class="chakra-icon css-13otjrl"
            data-testid="eye-on"
            fill="none"
            focusable="false"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
